// application/router/UrlRouting.ts
import {Routing} from "../Routing";
import {MakeUrlCommandSystem} from "application/system/UrlCommandSystem";
import {MakeUrlQuerySystem} from "application/system/UrlQuerySystem";


export default class UrlRouting implements Routing {

    public handlers() {
        return {
            getList: (req: any, res: any) => MakeUrlQuerySystem().getList(),
            getUrl: (req: any, res: any) => {
                const encode = MakeUrlQuerySystem().getUrl(req.params.encoded_id)
                res.writeHead(302, {
                    'Location': encode.toJSON().link
                });
                res.end();
            },
            makeUrl: (req: any, res: any) => {                
                return MakeUrlCommandSystem().createUrl({link: req.body.link})
            },
        };
    }

    public fillRoutes(router: any, handlers: any) {
        router.get("/list", handlers.getList);
        router.post("/make-url", handlers.makeUrl);
        router.get("/:encoded_id", handlers.getUrl);        
    }
}
